## Dataset Description

### Dataset Overview

The dataset used for this research, "Retail Store Sales - Dirty for Data Cleaning", simulates a point-of-sale transaction in a retail store. Specifically, it was made for practicing data analysis and cleaning. Several weaknesses that are frequently present in real-world business datasets, including missing values, inconsistent formats, and potential duplication, are included in the data, which is constructed to reflect actual retail transactions. 

### Data Source and Collection

This dataset is publicly available on Kaggle, created by Ahmed Mohammed. It was formulated to represent the real world complex nature of transactional sales records observed in retail settings. For instructional and analytical reasons, the dataset is structured to mimic realistic scenarios, even if it is not based on the records of a real business.

### Potential Implications of Data Collection

The dataset is simulated, thus it is free of the sampling biases and privacy issues that are usually present in real-world data.  But before any substantial analysis can occur, its purposeful roughness motivates students to use the proper data cleaning techniques.  The difficulties this dataset poses are reflective of real data management issues that arise in real life, even if the insights it offers might not be directly applicable to actual businesses.

### Structure of Data

The dataset is presented in a tabular format, where: 

- Rows represents a unique sales transaction
- Columns represents an important attribute of the transaction (e.g., item, quantity, payment method, etc.)
- The dataset contains 12,575 observations.

### Attribute Description

- Transaction ID - A unique identifier for each transaction. 
- Customer ID - A unique identifier for each costumer. 
- Category - The category of the purchased item. 
- Item - The name of the purchased item. 
- Price Per Unit - The statistic price of a single unit of the item. 
- Quantity - The quantity of the item purchased. 
- Total Spent - The total amount spent on the transaction. 
- Payment Method - The method of payment used. 
- Location - The location where the transaction occured. 
- Transaction Date - The date of the transaction. 
- Discount Applied - Whether discount is applied or not. 

## Data Cleaning

Before any Exploratory Data Analysis can be done to answer our research questions with the dataset, we have to clean the data first in case there are any issues that may result in problems with the analysis.

### Importing, Loading and Reading the Dataset

We begin by importing all the necessary libraries before starting the actual Data Cleaning process.

In [187]:
import numpy as np
import pandas as pd

Next, we load the dataset and call the `.head()` function to view a snippet of the dataset's contents.

In [188]:
df = pd.read_csv('retail_store_sales.csv')
df.head()

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location Transaction Date  \
0      10.0        185.0  Digital Wallet   Online       2024-04-08   
1       9.0        261.0  Digital Wallet   Online       2023-07-23   
2       2.0         43.0     Credit Card   Online       2022-10-05   
3       9.0        247.5     Credit Card   Online       2022-05-07   
4       7.0         87.5  Digital Wallet   Online       2022-10-02   

  Discount Applied  
0             True  
1             True  
2            False  
3              NaN  
4            False

Then the `.info()` function to view general information about the dataset.

In [189]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


This tells us that the dataset contains 12575 entries, but more importantly, we can quickly identify (3) potential issues we'll need to resolve as we go through the Data Cleaning process:

1. The `Discount Applied` column uses a Dtype of `object` rather than `boolean`.

2. The `Transaction Date` column uses a Dtype of `object` rather than `datetime64[ns]`. But since it also contains day, month, and year in one column, we can split these into their own respective Dtype `int64` columns.

3. The columns `Item`, `Price Per Unit`, `Quantity`, `Total Spent`, and `Discount Applied` all contain null values. Thus, we'll need to look into any inconsistencies or issues with these columns.

But before investigating these specific dataset issues, we'll start off with some preliminary checks by looking for duplicate values and duplicate categorical representations.

### Duplicate Values

In [190]:
df.duplicated().sum()

0

This dataset contains no duplicate values, so **no issues** there.

### Multiple Representations of the Same Categorical Value

The columns to watch out for here are `Category`, `Payment Method`, and `Location`. We have to ensure none of their representations end up being the same as other ones.

We'll start with `Category`.

In [191]:
df['Category'].unique()

array(['Patisserie', 'Milk Products', 'Butchers', 'Beverages', 'Food',
       'Furniture', 'Electric household essentials',
       'Computers and electric accessories'], dtype=object)

Next, `Payment Method`.

In [192]:
df['Payment Method'].unique()

array(['Digital Wallet', 'Credit Card', 'Cash'], dtype=object)

Lastly, `Location`.

In [193]:
df['Location'].unique()

array(['Online', 'In-store'], dtype=object)

Since none of the categories were representing the same thing as other categories, there are **no issues** to be found here.

### Incorrect Datatypes

For this portion, we'll resolve each of the previously mentioned columns individually.

#### Transaction Date Column

In order to make extracting the date into three separate columns easier, we'll first turn the `Transaction Date` into Dtype `datetime64[ns]` rather than its current Dtype of `object`. 

In [194]:
df['Transaction Date'] = pd.to_datetime(df['Transaction Date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12575 non-null  object        
 1   Customer ID       12575 non-null  object        
 2   Category          12575 non-null  object        
 3   Item              11362 non-null  object        
 4   Price Per Unit    11966 non-null  float64       
 5   Quantity          11971 non-null  float64       
 6   Total Spent       11971 non-null  float64       
 7   Payment Method    12575 non-null  object        
 8   Location          12575 non-null  object        
 9   Transaction Date  12575 non-null  datetime64[ns]
 10  Discount Applied  8376 non-null   object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 1.1+ MB


Afterwards, we may use `.dt.year`, `.dt.month`, and `.dt.day` to extract each part of the date and put them into their own separate columns.

In [195]:
df['Year'] = df['Transaction Date'].dt.year
df['Month'] = df['Transaction Date'].dt.month
df['Day'] = df['Transaction Date'].dt.day
df.head()

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location Transaction Date  \
0      10.0        185.0  Digital Wallet   Online       2024-04-08   
1       9.0        261.0  Digital Wallet   Online       2023-07-23   
2       2.0         43.0     Credit Card   Online       2022-10-05   
3       9.0        247.5     Credit Card   Online       2022-05-07   
4       7.0         87.5  Digital Wallet   Online       2022-10-02   

  Discount Applied  Year  Month  Day  
0             True  2024      4    8  
1             True  2023      7   23  
2            False  2022     10    5  
3              NaN  2022      5    7  
4            False  2022     10    2

Now that we've created columns `Year`, `Month`, and `Day` we no longer need the `Transaction Date` column, so we'll drop it.

In [196]:
df = df.drop(columns = 'Transaction Date')
df.head()

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location Discount Applied  Year  \
0      10.0        185.0  Digital Wallet   Online             True  2024   
1       9.0        261.0  Digital Wallet   Online             True  2023   
2       2.0         43.0     Credit Card   Online            False  2022   
3       9.0        247.5     Credit Card   Online              NaN  2022   
4       7.0         87.5  Digital Wallet   Online            False  2022   

   Month  Day  
0      4    8  
1      7   23  
2     10    5  
3      5    7  
4     10    2

#### Discount Applied Column

To turn the `Discount Applied` column's Dtype into `boolean` rather than its current Dtype of `object`, we'll use `.astype('boolean')`. 

In [197]:
df['Discount Applied'] = df['Discount Applied'].astype('boolean')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Discount Applied  8376 non-null   boolean
 10  Year              12575 non-null  int32  
 11  Month             12575 non-null  int32  
 12  Day               12575 non-null  int32  
dtypes: boolean(1), float64(3), int32(3), object(6)
memory usage: 1.0+ MB


Confirm the values using `.head()`.

In [198]:
df.head()

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location  Discount Applied  Year  \
0      10.0        185.0  Digital Wallet   Online              True  2024   
1       9.0        261.0  Digital Wallet   Online              True  2023   
2       2.0         43.0     Credit Card   Online             False  2022   
3       9.0        247.5     Credit Card   Online              <NA>  2022   
4       7.0         87.5  Digital Wallet   Online             False  2022   

   Month  Day  
0      4    8  
1      7   23  
2     10    5  
3      5    7  
4     10    2

### Missing Data

For this portion, we'll resolve each of the previously mentioned columns individually.

#### Item Column

One simple solution would be to drop all rows containing null values in the Item column. However, let's take a quick glance at the `Category` and `Item` columns.

In [199]:
df[['Category', 'Item']].head(8)

Category          Item
0     Patisserie   Item_10_PAT
1  Milk Products  Item_17_MILK
2       Butchers   Item_12_BUT
3      Beverages   Item_16_BEV
4           Food   Item_6_FOOD
5     Patisserie           NaN
6           Food   Item_1_FOOD
7      Furniture           NaN

Each item has the pattern `Item_##_CATEGORYID` with the category identifier being based on the value in `Category`. So while we can't figure out what item number was chosen, we can still figure out the category identifier for each item. After all, the `Category` column contains 0 null values.

So rather than dropping the null values for `Item` and losing potentially precious data, we can instead fill up all the rows for `Item` using only the category identifier.

We'll first create a dictionary of key-value pairs, with the content of `Category` serving as our key and the category identifier as the value.

In [200]:
category_id = {'Patisserie':'Item_NA_PAT', 'Milk Products':'Item_NA_MILK', 'Butchers':'Item_NA_BUT', 'Beverages':'Item_NA_BEV',
               'Food':'Item_NA_FOOD', 'Furniture':'Item_NA_FUR', 'Electric household essentials':'Item_NA_EHE',
               'Computers and electric accessories':'Item_NA_CEA'}
category_id

{'Patisserie': 'Item_NA_PAT',
 'Milk Products': 'Item_NA_MILK',
 'Butchers': 'Item_NA_BUT',
 'Beverages': 'Item_NA_BEV',
 'Food': 'Item_NA_FOOD',
 'Furniture': 'Item_NA_FUR',
 'Electric household essentials': 'Item_NA_EHE',
 'Computers and electric accessories': 'Item_NA_CEA'}

In [201]:
df['Item'] = df['Item'].fillna(df['Category'].map(category_id))
df[['Category', 'Item']].head(8)

Category          Item
0     Patisserie   Item_10_PAT
1  Milk Products  Item_17_MILK
2       Butchers   Item_12_BUT
3      Beverages   Item_16_BEV
4           Food   Item_6_FOOD
5     Patisserie   Item_NA_PAT
6           Food   Item_1_FOOD
7      Furniture   Item_NA_FUR

And if we run `.info()` on our dataset, `Item` should no longer contain any null values.

In [202]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              12575 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Discount Applied  8376 non-null   boolean
 10  Year              12575 non-null  int32  
 11  Month             12575 non-null  int32  
 12  Day               12575 non-null  int32  
dtypes: boolean(1), float64(3), int32(3), object(6)
memory usage: 1.0+ MB


#### Price Per Unit Column

Filling up the null values for `Price Per Unit` is doable as long as both `Quantity` and `Total Spent` contain values. All we have to do is compute what the price per unit would be based on the quantity and total spent.

In [203]:
df['Price Per Unit'] = df['Price Per Unit'].fillna(df['Total Spent']/df['Quantity'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              12575 non-null  object 
 4   Price Per Unit    12575 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Discount Applied  8376 non-null   boolean
 10  Year              12575 non-null  int32  
 11  Month             12575 non-null  int32  
 12  Day               12575 non-null  int32  
dtypes: boolean(1), float64(3), int32(3), object(6)
memory usage: 1.0+ MB


#### Quantity and Total Spent Columns

Suspiciously enough, both `Quanity` and `Total Spent` contain the same number of null values. We'll investigate them first to see if they both share null values.

In [204]:
(df['Quantity'].isna() == df['Total Spent'].isna()).all()

True

While we were able to fill up `Price Per Unit` using `Quantity` and `Total Spent` as reference, the opposite doesn't hold true. Since we can't fill these missing values, we'll delete all rows containing null values in both columns

In [205]:
df.dropna(subset = ['Quantity', 'Total Spent'], inplace = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11971 entries, 0 to 12574
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    11971 non-null  object 
 1   Customer ID       11971 non-null  object 
 2   Category          11971 non-null  object 
 3   Item              11971 non-null  object 
 4   Price Per Unit    11971 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    11971 non-null  object 
 8   Location          11971 non-null  object 
 9   Discount Applied  7983 non-null   boolean
 10  Year              11971 non-null  int32  
 11  Month             11971 non-null  int32  
 12  Day               11971 non-null  int32  
dtypes: boolean(1), float64(3), int32(3), object(6)
memory usage: 1.1+ MB


#### Discount Applied Column

Removing the null values in `Discount Applied` would be a huge loss in data, considering how many rows would be removed.

In [206]:
# All columns have the same number of non-null values except for Discount Applied
df['Item'].count() - df['Discount Applied'].count()

3988

As you can see, we would lose `3988` rows of data. Instead of dropping the rows with null values, let's instead calculate for the mode. We use mode here over mean and median because the `Discount Applied` column contains categorical data.

In [207]:
df['Discount Applied'] = df['Discount Applied'].fillna(df['Discount Applied'].mode()[0])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11971 entries, 0 to 12574
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    11971 non-null  object 
 1   Customer ID       11971 non-null  object 
 2   Category          11971 non-null  object 
 3   Item              11971 non-null  object 
 4   Price Per Unit    11971 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    11971 non-null  object 
 8   Location          11971 non-null  object 
 9   Discount Applied  11971 non-null  boolean
 10  Year              11971 non-null  int32  
 11  Month             11971 non-null  int32  
 12  Day               11971 non-null  int32  
dtypes: boolean(1), float64(3), int32(3), object(6)
memory usage: 1.1+ MB


### Inconsistent Formatting

In [208]:
df.head()

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location  Discount Applied  Year  \
0      10.0        185.0  Digital Wallet   Online              True  2024   
1       9.0        261.0  Digital Wallet   Online              True  2023   
2       2.0         43.0     Credit Card   Online             False  2022   
3       9.0        247.5     Credit Card   Online              True  2022   
4       7.0         87.5  Digital Wallet   Online             False  2022   

   Month  Day  
0      4    8  
1      7   23  
2     10    5  
3      5    7  
4     10    2

Columns `Transaction ID`, `Customer ID`, and `Item` all follow a specific formatting. Let's make sure there are 0 inconsistencies and that this format is followed throughout the entire dataset by using regex.

In [209]:
# Starts with TXN_ and ends with 7 digits
txn_regex = r"^TXN_\d{7}$"
invalid_txn = df[~df['Transaction ID'].str.match(txn_regex)]

invalid_txn.empty

True

In [210]:
# Starts with CUST_ and ends with 2 digits
cust_regex = r"^CUST_\d{2}$"
invalid_cust = df[~df['Customer ID'].str.match(cust_regex)]

invalid_cust.empty

True

In [211]:
# Starts with Item_, then 1 or 2 letters/digits, and ends with 3 or 4 letters
item_regex = r"^Item_\w{1,2}_[A-Z]{3,4}$"
invalid_item = df[~df['Item'].str.match(item_regex)]

invalid_item.empty

True

There are zero inconsistencies with the formatting, so **no issues** here.

## Research Questions & Exploratory Data Analysis